---
title: "Read VCF with vcfR"
author: "Alexey Larionov"
date: "10 Apr 2021"
output: 
  html_document:
    toc: true
    toc_float: true
    number_sections: true
editor_options: 
  chunk_output_type: console
---

# Summary

- Import VCF file into vcfR object  
- Extract header and variants data  
- Extract genotype data: GT, DP, GQ, AF and AD  

Requires 60GB RAM instance (e.g. r4.2xLarge or m4.4xlarge)  
Requires ~1GB storage space for the RData file that will be created  

<style>
pre{
  overflow-x: auto;
}
pre code{
  word-wrap: normal;
  white-space: pre;
}
s</style>

```{r echo=F}
options(width=999)
```

# Start section

```{r}

# Time
Sys.time()

# Memory info
gc()

# Clean up
rm(list=ls())
graphics.off()

# Options
options(stringsAsFactors = F)

# Working folders
base_folder <- "/home/share"
data_folder <- file.path(base_folder,"data/s05_read_to_R")
#dir.create(data_folder)
scripts_folder <- file.path(base_folder,"scripts/s05_read_to_R")
setwd(scripts_folder)

# Required library
#install.packages("Rcpp")
#install.packages("vcfR")
library(vcfR)

```

# Read VCF

```{r}

# VCF file name
vcf_file <- file.path(base_folder,"data","s04_annotate","IHCAPX8_dragen_joint.hard-filtered.PF.MA-flag.MA-split.ID.ClinVar.std-Chr.Reheaded.VEP.split-VEP.vcf.gz")

# Read vcf to vcfR object 
data.vcfr <- read.vcfR(vcf_file, verbose = F) # verbose = F for not to clutter the log

# Check result
data.vcfr
head(data.vcfr)

# Clean-up
rm(vcf_file)

```

# Extract header and variants information to data frames 

Takes ~2hrs and requires ~52GB RAM

```{r}

# Get list with data from VCF header and fixed/standard VCF columns
meta_fix <- vcfR2tidy(data.vcfr, info_only=T) # dot_is_NA = TRUE ??

# Get data frame with meta-information from vcf header
vcf_header.df <- meta_fix$meta
dim(vcf_header.df)
colnames(vcf_header.df)
head(vcf_header.df)
tail(vcf_header.df)

# Get data frame with variants information
variants.df <- as.data.frame(meta_fix$fix)
dim(variants.df)
str(variants.df,list.len=Inf) # dot_is_NA = TRUE didnt work for vcfR2tidy

# Clean-up
rm(meta_fix)
gc()

```

# Extract genotype data: GT, DP, GQ and AD  

```{r}

# Extract genotypes as numeric codes: 0/1, 1/1 etc 
gt.mx <- extract.gt(data.vcfr) # , element="GT" by default
dim(gt.mx)
gt.mx[1:5,1:3]
sum(is.na(gt.mx))/(ncol(gt.mx)*nrow(gt.mx))

# Extract genotype depth (its a numeric matrix)
dp.mx <- extract.gt(data.vcfr, element="DP", as.numeric = T)
dim(dp.mx)
dp.mx[1:5,1:3]
sum(is.na(dp.mx))/(nrow(dp.mx)*ncol(dp.mx))
quantile(dp.mx, na.rm=T)
hist(dp.mx) 
hist(log10(dp.mx))
hist(dp.mx[dp.mx<100]) 
abline(v=10, lty=2, col="red")
sum(dp.mx < 10, na.rm=T)/(nrow(dp.mx)*ncol(dp.mx)) # fraction of genotypes to be removed because of low DP

# Extract genotype quality (its a numeric matrix)
gq.mx <- extract.gt(data.vcfr, element="GQ", as.numeric = T)
dim(gq.mx)
gq.mx[1:5,1:3]
sum(is.na(gq.mx))/(nrow(gq.mx)*ncol(gq.mx))
quantile(gq.mx, na.rm=T)
hist(gq.mx)
abline(v=20,lty=2,col="red")
sum(gq.mx < 20, na.rm=T)/(nrow(gq.mx)*ncol(gq.mx)) # fraction of genotypes to be removed because of low GQ

# Extract allele fraction (its a numeric matrix)
af.mx <- extract.gt(data.vcfr, element="AF", as.numeric = T)
dim(af.mx)
af.mx[1:5,1:3]
sum(is.na(af.mx))/(nrow(af.mx)*ncol(af.mx))
quantile(af.mx, na.rm=T)
hist(af.mx)
abline(v=0.1,lty=2,col="red")
abline(v=0.9,lty=2,col="red")

# Extract allelic depth
ad.mx <- extract.gt(data.vcfr, element="AD")
dim(ad.mx)
ad.mx[1:5,1:3]
sum(is.na(ad.mx))
sum(ad.mx == ".,.", na.rm=T)
sum(ad.mx == "0,0", na.rm=T)
sum(ad.mx == "0,0", na.rm=T)/(nrow(ad.mx)*ncol(ad.mx))

# Clean-up
rm(data.vcfr)
gc()

```

# Summarize and save

```{r}

ls()

dim(vcf_header.df)
dim(variants.df)

dim(dp.mx)
dim(gq.mx)
dim(gt.mx)
dim(af.mx)
dim(ad.mx)

save.image(file.path(data_folder,"s01_import_vcf_to_R.RData"))

```

# Final section

```{r}

sessionInfo()
gc()
Sys.time()

```
